lib/ndr_import/file/xml.rb

```diff
@@ -16,32 +16,87 @@ def initialize(*)
         super
 
         @pattern_match_xpath = @options['pattern_match_record_xpath']
+        @xml_file_metadata   = @options['xml_file_metadata']
+        @options['slurp'] ? prepare_slurped_file : prepare_streamed_file
       end
 
       private
 
+      def prepare_slurped_file
+        @doc = read_xml_file(@filename)
+        slurp_metadata_values
+      end
+
+      def prepare_streamed_file
+        with_encoding_check(@filename) do |stream, encoding|
+          @stream   = stream
+          @encosing = encoding
+        end
+        stream_metadata_values
+      end
+
+      def slurp_metadata_values
+        return unless @xml_file_metadata.is_a?(Hash)
+
+        self.file_metadata = @xml_file_metadata.transform_values do |xpath|
+          @doc.xpath(xpath).inner_text
+        end
+      end
+
+      def stream_metadata_values
+        return unless @xml_file_metadata.is_a?(Hash)
+
+        self.file_metadata = @xml_file_metadata.transform_values.with_index do |xpath, index|
+          # Ensure we're at the start of the stream each time
+          @stream.rewind unless index.zero?
```

Comment:
You can't always rewind a stream (here and later in this same file).

I don't know whether this affects any of our code, e.g. reading data from within zip files. I don't think this causes breakage to existing code, because `xml_file_metadata` will be `nil`.

I can't see any other calls to `rewind` (or `seek(0)`) except the ones you've added. But presumably the stream always comes from `with_encoding_check`, which opens a `SafePath`, which would usually be rewindable (we don't generally use pipes or fifes). So maybe it's not a change in stream requirements?

Reply:
Thanks @bshand - yes, I considered this and agree that I think it's fine. The streaming logic is only in the Xml file handler (so wouldn't be used on any other file types) and the stream is only produced via `with_encoding_check`.